frontend: Use system locale with UTF-8 instead of 'C'

Dankirk opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description

The changes here are based on PR #13097 that was split from this issue and does some preliminary work to support changes introduced here.

- Sets C runtime locale to system locale with UTF-8 codepage on Windows.
This has always been [default behavior on unix](https://github.com/qt/qtbase/blob/dev/src/corelib/kernel/qcoreapplication.cpp#L569), but Windows defaults to minimal 'C' locale.
  - `LC_NUMERIC` is still set to `"C"`. Now on all platforms instead of just unix. 
This is so decimal point is a dot (not a comma) for string <-> float conversions.

- The configured CRT locale is copied to be the default `std::locale` for C++. 
All platforms have been using minimal `"C"` until now. This change affects new `facet` and `ios_base` instances without a specified locale or `imbue()` call.

- OBS Studio language setting no longer changes `QLocale`'s default locale and instead always uses system locale.
This gives conformity with non Qt functions, but most importantly is likely what user wants as well. Ie. sorting and formatting functions should follow OS locale rules instead of OBS Studio translations language. (Reverts c4840ddba05a8c2d01568445645097704595ea47)

- `obs_get_locale()` still returns OBS language locale, which is used for Python and LUA apis, GDI+ text widget transformations, and HTTP accepted languages header. 

<!--- Describe your changes in detail. -->
<!--- If this change includes UI elements, please include screenshots. -->

### Motivation and Context

Locale-aware operations like sorting and time formatting in C are not available on Windows, but are on unix, as pointed out in PR #12577. 
Fixes #11133, fixes #12953

The C++ locale and QLocale changes make the locale-aware functions of all layers work in similiar fashion. 

For example: On unix currently the used locales are: OS locale for CRT, minimal `"C"` for C++ and OBS language for QLocale.
A weekday name can be in three different languages depending if you used `strftime()`, `std::time_get` facet or `QLocale`.
This makes string transformations between C, C++ and Qt very tricky.


<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open GitHub Issue, or implements feature request -->
<!--- from the Ideas page, please link to the issue here. -->

### How Has This Been Tested?

An important point is that the CRT locale settings introduced here have always been this way for unix, which suggests there aren't any insurmountable problems with the new locales. Windows specific functions should be tested for CRT locale. Changes for C++ and QLocale defaults affect all platforms.

Searched the codebase for affected areas and addressed as necessary:
- [x] CRT:  [_ctype.h_](https://en.cppreference.com/w/c/header/ctype.html) character classification function parameters and expected return values
- [x] CRT: `strftime()` formatting with `%` placeholders
- [x] CRT: `scanf()` and `printf()` formatting with `%` placeholders
- [x] CRT: `FILE` operations
- [x] C++: `fstream` operations
- [x] C++: `facet` locale usage
- [x] QLocale: Expected return values of formatting functions
- [x] QLocale: `QString` locale-aware methods


Some general testing with Japanese characters
- Edited recording path with %A (weekday) variable and some Japanese characters. Recorded a video. Weekday name was localized and recording worked fine.
- Remuxed said file. Worked fine.
- Renamed some sources with Japanese characters and exported the scene collection, removed it from OBS and re-imported it. No problems.
- Wrote names of those sources to logFile with blog()


I'm on Windows 11 English US version, but with Finnish locale settings (fi_FI). OBS language is English.

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment (hardware, OS version, etc.),-->
<!--- and the tests you ran, including how it may affect other areas of code. -->

### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
<!--- - Bug fix (non-breaking change which fixes an issue) -->
- New feature (non-breaking change which adds functionality)
<!-- - Tweak (non-breaking change to improve existing functionality) -->
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
- Breaking change (fix or feature that would cause existing functionality to change)
  -  3rd party Python, LUA and rtmp that have been using `_mbs_` conversion functions directly or via file io operations have to input text in utf-8 and expect utf-8 output (except for `wchar`/`_wcs_` which is OS defined). 

<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

Dankirk committed:
frontend: Use system locale instead of 'C'

Sets runtime locale to system locale with UTF-8 codepage. This is already default behavior on unix, but Windows defaults to minimal 'C' locale.

Use CRT locale for C++ std::locale default

OBS Studio language settings no longer change QLocale default locale, instead system locale is used for conformity. It is likely this is what user wants as well. Ie. sorting and formatting functions should follow OS locale instead of OBS Studio language (which also lacks country information).

Dankirk commented:
Scouted the web and the codebase for potential issues. Here's some observations...

[General stuff about setlocale()](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/setlocale-wsetlocale?view=msvc-170) on Windows
For list of things C runtime locale affects https://cplusplus.com/reference/clocale/
- We can ignore all number related things, because we use the minimal 'C' locale for `LC_NUMERIC`.
- Affected _string.h_ and _time.h_ functions are all functions that are specifically for locale-aware things, Things like weekday names will now be localized using `strftime()`.
- [_ctype.h_](https://cplusplus.com/reference/cctype/) character classification ranges are extended. ie. [`isalnum()`](https://cplusplus.com/reference/cctype/isalnum/) may return true for more characters, so there is reason to check if any part using these functions should be specifically ASCII range. Couldn't hurt to cast the parameters to `unsigned char` either, since many functions expected value to be 0-256, which `char` casted to `int` might not be. Then again, the functions in use have worked fine on unix until now...
- [_stdio.h_ ](https://cplusplus.com/reference/cstdio/) TODO: file operations need to be looked at.


### multibyte <-> utf8 <-> wchar
In _platform.h_ there are various string conversion functions. From these only the multibyte functions with `_mbs_` are affected by this change. The rest use Windows API, which doesn't follow C runtime locale modified by `setlocale()`. 

The `_mbs_` functions are currently not used in OBS Studio itself, but are offered for external usage for Python, LUA and rtmp. This means there's no change for OBS Studio itself, but external things might see different results from these conversion functions on Windows, which will now be more alike to return values on unix.

The utf8 <-> wchar functions (ie `os_utf8_to_wcs()`) use `MultiByteToWideChar()` and `WideCharToMultiByte()` functions with utf-8 codepage, which will after this update. Unlike `mbtowcs()` and the sort in  `_mbs_` implementations, these functions are independent from CRT locale.


### When to setlocale() ?
The beginning of `OBSApp `constructor seems like a good place for `setlocale()` calls, since it is the earliest after Qt has performed it's own locale calls. Unix has also been using the constructor to reset `LC_NUMERIC` back to C. If it is decided that OBS translations locale should be followed instead of OS's, `initLocale()` also seems acceptable. 

It is important runtime locale is set as soon as possible, so there are no active io streams that would get weird results after the change. Any `blog()` or `fprint()` before `setlocale()` would also mean the underlying streams have to be reset in a way or another or they should lead to programs termination. Notably `fstream logfile;` is declared before this point, but is yet to open anything. To me the current state looks good.

Dankirk committed:
frontend,libobs,plugins: Ensure ctype functions use unsigned chars

Cast ctype function char parameters to unsigned char to ensure they are in correct range (0 to 255 vs -128 to 127) when used with utf-8 encoding (or extended ascii).

Fixes dstr astrcmp* functions when used with utf-8 (or extended ascii) characters, so now they are treated greater than the base ascii and thus sorted after them, not before.

Dankirk committed:
logging: Use ISO 8601 time format for logging

Switch locale-aware timestamping for logging / crash handling to %H:%M:%S

Update frontend/OBSApp.cpp

Co-authored-by: Patrick Heyer <PatTheMav@users.noreply.github.com>

Update libobs/obs-win-crash-handler.c

Co-authored-by: Patrick Heyer <PatTheMav@users.noreply.github.com>

Fenrirthviti commented:
My main concern here, as someone who only uses the English/USA locale/region, is that I'm unsure what the expectation for a Windows application is. The current motivation seems to be "Unix does it this way" and that to me, is not sufficient. Do we have examples and recommendations from Microsoft, or other prominent Windows applications on how they handle this kind of setting for apps that use translations?

Dankirk commented:
My own take is that OS regional settings + app translations is the desired output with no additional settings in UI, good likelyhood being fine by default, but allows configuration when needed. The obvious drawback is that to change the regional settings one needs to change OS settings, which could be an issue when using a shared device, but OS should provide options for it.

Many Microsoft apps understandably just follow the OS region settings. That includes the file explorer. Apps like Office do offer a separate in app setting for regional settings as well, but that's because they specialize in that sort of thing. For apps in general many have translation + maybe time formatting options and don't follow a specific region per se. The sorting order is a gamble. Steam seems to sort games and friends using the app display language. Spotify sorts playlists by Windows display language (not region settings, nor Spotify display language, I don't recommend this). Whatsapp uses OS regional settings.

In any case, any locale is still better than the current situation with non-changeable "C" locale.

Fenrirthviti commented:
Thanks for the additional context here. My, admittedly mostly uninformed opinion based on the discussion here, is that this seems fine. Without lack of a clear "best practice" on Windows, moving things in-line with our cross-platform implementation seems like the best option.

@PatTheMav or @jcm93 Does macOS follow a similar approach to what is being proposed here?

Dankirk commented:
The obs.manifest changes has been added and locale setup routine moved to beginning of the app, with unix re-running it after OBSApp has been created. 

If there's any issues with locale setup routine being the first thing to do please let me know. 
The reasoning for it is that many things inherit the locale when they are initialized and retroactively updating that is cumbersome. This includes things like outputting to anything to `stdout` or using `fstream`. On my observation `load_debug_privilege()` is the first function to use `blog()`, so the locale setup routine should happen before that.

PatTheMav commented:
The main "problem" this PR has to address is that OBS on the whole is conceptually not built to be aware (much less so capable of handling) of locales (and locale differences), similarly to how it pretends that any set of bytes in memory is "UTF-8" but then happily treats it as "just ASCII". 

For better or worse all that OBS can handle without issue is US-American text and formats, anything beyond that is a "happy accident".

Introducing these capabilities has to go _far beyond_ just slapping on some `setlocale()` calls or adding facets where necessary, because of the following aspects:

* C only supports _global_ locales that indeed have to be set almost immediately at an app's run time to ensure all system APIs follow suit
* The global locale is set _per process_ on POSIX, but _per thread_ on Windows, which makes locale-awareness trickier to ensure for all the threads `libobs` uses in the latter case  (and that's not even getting into the weeds of whether certain threads need to be locale aware and _if_ they suddenly become so - because of the global switch on POSIX - we introduce hard to debug/notice side-effects).
    * And many (most?) Windows C APIs do not even respect `setlocale`?
* C++ supports setting up custom locale objects and allows to apply them where needed, thus allowing the app to be much more "locale-aware" (e.g. one can use different locales for "reading" and "writing" of data, at least for locale-aware library functions).
* As if two (or three in the case of Windows) different layers of locale handling are not enough we also use Qt on top of it all, which seems to follow a similar approach like C++ (with `QLocale` instances)
* Note that C++'s and Qt's approach is considered "correct" by requiring the app to _explicitly_ use locale awareness where needed instead of flipping a "magic switch" that changes everything behind the scenes. Because it makes sense for an app to use an internal "canonical" format, but use local-awareness when interfacing with the user and the user's data.

That's already a big pile of work to get through and figure out where/how/if to change OBS (and `libobs`, and the 1st party modules) to become locale-aware, but on top of that we have the (largely non-existent) Unicode-handling:

* Qt's `QString`s are all more or less expected to be encoded in UTF-16 (I haven't checked if it's actually UTF-16 or just UCS-2 with UTF-16 surrogate pairs)
* `std::string` instances and `char *` data is ostensibly thought of as "UTF-8" but indeed is treated as plain ASCII or simply a sequence of bytes in memory.
* As long as such a "string" is not truncated, manipulated, or collated (and the unchanged array of bytes copied/transferred as-is) the underlying strings will be retained
    * Because those strings are passed as-is to Qt and Qt is indeed UTF-8 aware, the UTF-8 code points will be correctly decoded and re-encoded as UTF-16 when stored in a `QString`
* On Windows any `wchar_t`-based string is considered to be UTF-16 as well (in the case of file system paths that might actually still be UCS-2), which is then passed to Windows APIs to convert to/from UTF-8 (similar to Qt)
* Almost all _internal_ APIs (or functions) that handle `char *` data or `std::string` do _not_ support UTF-8: All bytes are considered "ASCII characters", there is no capability to detect the variable-length encoding of UTF-8 code points, much less so any capability to detect grapheme clusters, or any awareness of language-specific differences in meaning of what "upper case" means.
    * Again, as long as no manipulation or interpretation of the strings take place this is fine, because UTF-8 is then just a "transport" format.
* As mentioned in the description, on Windows there is the additional complexity of many APIs being available as an "ANSI" variant (which uses the current code page for encoding/decoding) and the "UNICODE" variants (which might either be actual variable-length UTF-16 or fixed-length UCS-2).
    * And as if _all of that_ weren't bad enough, Microsoft added the mentioned support for a "UTF-8 code page" which allows one to use UTF-8 encoded text with the ANSI variants of APIs, which in theory requires developers to un-adopt the use of their UNICODE APIs. 

---

Mind you, I'm not saying that we shouldn't adopt changes like these, but that these changes require a great deal of thought and need to address many architectural design flaws in OBS as it is right now. Simply adding it in bits and pieces runs the risk of fixing symptoms but not fixing the core issue(s), particularly as the entire app has not been designed to properly handle locale's or anything beyond ASCII text and making it handle text "right" in one area might collapse a whole house of cards of assumptions about character data in another.

For that (and a few other reasons) it might take same time (and might even require splitting the whole endeavour up into separate "units") to get it over the finish line.

Dankirk commented:
The Qt locale setup seems to default to `""`, which sets the default system locale and also default encoding. Unix is mainly utf-8 default so this works for most setups I believe. I wanted to be sure though, and enforce UTF-8. The code itself has a [bunch of lines](https://github.com/qt/qtbase/blob/dev/src/corelib/kernel/qcoreapplication.cpp#L577) like `// FIXME: Shouldn't we still setlocale("UTF-8")?` Which also contributed to the fact that I probably should. Otherwise the new locale routine follows Qt's idea of what localization should be, like falling back to `C.UTF-8`, so there's no harm done. Also on principle Qt naturally only does what is required by Qt anyway, which may not be what we wanted for C locale (as long as we are compatible). On the real though I admit it would be pretty hard to find a failing setup for the Qt's version, so for that it's also acceptable if we revert to just setting LC_NUMERIC to C for unix in the constructor. (We still need the std::locale:.global though)

For commandline, yes GetCommandLineW would work and also be better than the first workaround that came to my mind about acp->wide->utf8 conversion. I still think the manifest approach is somewhat preferable as an environment configuration rather than introducing more platform specific code. I also see changing of the -A apis by the active code page as thing that would help adoption of 3rd  party code. If they used -A apis they either didn't know better or maybe they maintained some existing code later called `MultibyteToWideChar()` with `CP_ACP` flag. Both of these cases is fixed/compatible by changing the ACP to UTF-8. Not saying we should convert from -W apis. We can keep that.

For time things, yes we should use a standardized format and use the tools that most approriate in context (Qt). It is somewhat out of the scope of this, so I settled with preserving current results/expectations as much as possible. This can be changed to do ISO formatting though.

PatTheMav commented on libobs/obs-win-crash-handler.c:
```diff
@@ -245,7 +245,7 @@ static inline void write_header(struct exception_handler_data *data)
 	struct tm ts;
 	ts = *localtime(&now);
-	strftime(date_time, sizeof(date_time), "%Y-%m-%d, %X", &ts);
+	strftime(date_time, sizeof(date_time), "%Y-%m-%d, %T", &ts);
```
```suggestion
	strftime(date_time, sizeof(date_time), "%F, %T", &ts);
```
See above.

PatTheMav reviewed (commented):
Explicit casts to `unsigned char` should be fine IMO, using the locale-independent time format strings is also more correct, though we could use the same corresponding fixed format for the date as well (as commented).

The other changes are still pending discussion.

PatTheMav commented:
I forgot to mention in my previous comment that I am still undecided whether to still use the UTF-8 codepage as a "stopgap" until we properly fix the command line parsing code though. It's still a kludge, but as there are plans to clean up that code anyway, it would be a temporary kludge we can live with.

PatTheMav commented:
Alright, slept on it, here's my suggestion:

Could you please package all changes of this PR _except_ for the changes related to `setlocale` into a separate PR?

Off the top of my head that should be:

* Adjustment of time format strings to `%F, %T`
* Explicit cast to `unsigned char`
* Use of UTF-8 code page for ANSI/C API methods

I'll probably want to have a comment in `obs-main.cpp` above `obs_set_cmdline_args` that explicitly mentions that this requires the UTF-8 code page to be set in the manifest.

I can then have someone else review those changes as well and we can discuss those in isolation from the larger issue of how/where to do locale-specific stuff in the app.

Dankirk committed:
frontend: Add utf-8 active code page manifest

Declaring Utf-8 as active code page in manifest makes Win32 API use utf-8 instead of ANSI codepages when using the "A" versions of functions.

Manifest declaration also encodes command line arguments as utf8. This allows for example --profile <name> to load profiles with special characters.

Dankirk commented:
> As I mentioned earlier, the core issue is that nothing in OBS' design is built to be aware of locale-specific changes to C APIs and indeed not to the finer points of UTF-8 encoded data

We still do not need everything to be locale or encoding aware. We can use them for what we have and just understand that they have that limitation about them. We only need them to not crash and be deterministic. It would still be good if there was something available to use when we do need the awareness.


> I also haven't checked whether all C library functions whose behaviour is influenced by the current locale have a corresponding variant that allows passing in a different locale. Otherwise, and because OBS is a multithreaded program, the "current locale" becomes a global state variable that will need to be "locked" similarly to the graphics context.

Plain C doesn't support passing locales, it (std::locale and facets) are a C++ feature for which the overcoming the earlier process/thread-basedness was one of the core things to address. It would work without locks, but needs features outside of C.

> Actual locale-awareness is limited to two functions: os_strtod and os_dtostr.

A bit beside the point, but interestingly these functions currently and after do nothing different from standard `strtod` and `dtostr`. Because the numeric locale is and will be `'C'` on all platforms, which means decimal points are dots in all setups. Except for Qt since it's independent. If you did supply a string/double from Qt api to this function or vice versa, it will fail to convert the decimal point.

> And while that will allow functions like strcoll to do the right thing (e.g. putting "ch" between "h" and "i" in Czech), it will not be sufficient for converting characters in languages like Turkish. Even with the correct locale set, toupper and tolower will not be able to handle sinirli/SİNİRLİ (angry) and sınırlı/SINIRLI (limited). Lowercase "angry" becomes uppercase "limited", but their opposites become gibberish when not handled correctly ("S??N??RL??" in the case of sınırlı, which seems to suggest that the function took those bytes and just incremented their value which then created invalid UTF-8 code units).

Indeed using tolower/toupper (or strmpi for case insensitive comparison) is not sufficient for anything locale-aware outside of ascii. If locale awareness is needed, they should be done with strcoll (in which case sensitivity should probably be respected, because locale will handle the case folding), but in majority of cases where you want case-insensitivity you do it for ASCII things. 
You also need to think whether you actually need true locale / case sensitivity or is the current byte-wise desired. For example checking for existing profile or source name, because then you could change locale and affect the case sensitivity comparison to match/not match a profile / source name.
